master/trials.tx.up.sql

```diff
@@ -0,0 +1,76 @@
+    id integer PRIMARY KEY GENERATED BY DEFAULT AS IDENTITY,	
+    user_id integer REFERENCES public.users(id) ON DELETE CASCADE NOT NULL,
+    project_id integer REFERENCES public.projects(id) ON DELETE CASCADE NOT NULL,
+	  name text NOT NULL,
+	  filters jsonb NOT NULL,
+    sorter  jsonb NOT NULL,
+  CONSTRAINT project_filter_name_uniq UNIQUE (project_id, name)
+);
+
+-- make analogous modifications to those found in reindex-steps pr
+-- because we are joining validations on (trial_id, total_batches)
+CREATE INDEX validations_archived ON raw_validations(archived);
+ALTER TABLE raw_validations DROP CONSTRAINT validations_trial_id_run_id_total_batches_unique;
+CREATE UNIQUE INDEX validations_trial_id_total_batches_run_id_unique ON raw_validations(trial_id, total_batches, trial_run_id);
+
+
+CREATE AGGREGATE jsonb_collect(jsonb) (
+  SFUNC = 'jsonb_concat',
+  STYPE = jsonb,
+  INITCOND = '{}'
+);
+
+
+-- aggregations should be mostly redundant.the only way they 
+-- would get applied is when there are multiple trial_run_ids for a single 
+-- (trial_id, total_batches) on steps or validations. in which case,
+-- which trial_run_id is it appropriate to take the metrics from?
+-- there have been situations where they differ.
+CREATE OR REPLACE VIEW public.trials_augmented_view AS 
+  WITH b AS (
+    select trial_id, max(total_batches) total_batches from steps group by trial_id
+  )
+  SELECT
+      t.id AS trial_id,
+      t.state AS state,
+      t.hparams AS hparams,
+      jsonb_collect(s.metrics->'avg_metrics') AS training_metrics,
+      jsonb_collect(v.metrics->'validation_metrics') AS validation_metrics,
+      t.tags AS tags,
+      t.start_time AS start_time,
+      t.end_time AS end_time,
+      max(e.config->'searcher'->>'name') as searcher_type,
+      max(e.id) AS experiment_id,
+      max(e.config->>'name') AS experiment_name,
+      max(e.config->>'description') AS experiment_description,
+      -- there's only one
+      jsonb_agg(e.config ->> 'labels'::text) AS experiment_labels,
+      max(e.owner_id) AS user_id,
+      max(e.project_id) AS project_id,
+      max(p.workspace_id) AS workspace_id,
+      -- temporary
+      max(b.total_batches) as total_batches,
+      max(e.config->'searcher'->>'metric') AS searcher_metric,
+      max(v.metrics->'validation_metrics'->>(e.config->'searcher'->>'metric')) AS searcher_metric_value
+  FROM trials t
+  LEFT JOIN experiments e ON t.experiment_id = e.id
+  LEFT JOIN projects p ON e.project_id = p.id
+  LEFT JOIN validations v ON t.id = v.trial_id AND v.id = t.best_validation_id
+  LEFT JOIN steps s on t.id = s.trial_id AND v.total_batches = s.total_batches
+  LEFT JOIN b on t.id = b.trial_id 
+  group by t.id
+  -- find other subquery way to do this??
+
+
+-- is the assumption here valid? will we always have the row in steps for a corresponding row in validations?s
+-- does avg_metrics correspond to the actual state at that batch?
+-- or an average over previous batches?
```

Comment:
note here: "this assumption" refers to the assumption that we could use (paraphrasing) `batch = validations.batch where validation.id = trial.best_validation id` and basically get all the validation and training metrics from that same batch. 

the goal being for each metric to associate a value to a trial.

This assumption remains valid, [for now](https://determinedai.atlassian.net/browse/DET-4292?focusedCommentId=16689), but It is worth calling attention to the fact that this is _not ideal_. For example: searcher metric might be best at one point, but another validation metric is better at another point. 

Given the alternatives though:
- scanning raw_validations/steps
- doing a different "index" in some form or fashion for each validation metric

this seemed the least bad option. In any case searcher metric is probably a decent proxy for the other metrics, and they can drill into more how the various metrics evolved on the charts

Reply:
apparently`JSONB` indexes on fields with numeric values [can be used to efficiently query this kind of thing](https://stackoverflow.com/a/30122313/9778768), but it would be a pretty massive index probably.